internal/server/spanner/datasource.go

```
@@ -77,7 +77,27 @@ func (sds *SpannerDataSource) Node(ctx context.Context, req *pbv2.NodeRequest) (
 
 // Observation retrieves observation data from Spanner.
 func (sds *SpannerDataSource) Observation(ctx context.Context, req *pbv2.ObservationRequest) (*pbv2.ObservationResponse, error) {
-	variables, entities, entityExpr := req.Variable.Dcids, req.Entity.Dcids, req.Entity.Expression
+	if req.Entity == nil {
+		return nil, fmt.Errorf("entity must be specified")
+	}
+
+	entities, entityExpr := req.Entity.Dcids, req.Entity.Expression
+	if len(entities) > 0 && entityExpr != "" {
+		return nil, fmt.Errorf("only one of entity.dcids and entity.expression should be specified")
+	}
+
+	variables := []string{}
+	if req.Variable != nil {
+		// Variable expressions are not yet supported in Spanner.
+		if req.Variable.Expression != "" {
+			return nil, nil
```

Should we return an error instead of failing silently here?

---

This was so the whole request doesn't fail (for example remote data source supports variable expressions, so we could still have a valid response even if spanner data source doesn't return anything)

---

SG. Perhaps let's log a message in that case.